graphcore_cloud_tools/paperspace_utils/symlink_datasets_and_caches.py

```diff
@@ -282,26 +289,36 @@ def download_file(
     config = TransferConfig(max_concurrency=max_concurrency)
     target = Path(file.local_file)
     target.parent.mkdir(exist_ok=True, parents=True)
-    exception = None
-    try:
-        if not use_cli:
-            s3client.download_file(bucket_name, file.s3file, str(target), Config=config)
-        else:
-            cmd = (
-                f"aws s3 --endpoint-url {aws_endpoint} --profile {aws_credential} "
-                f"cp s3://{bucket_name}/{file.s3file}"
-                f" {target}"
-            ).split()
-            print(cmd)
-            out = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, text=True)
-            if out.returncode != 0:
-                raise S3DownloadFailed(f"Command {cmd} failed with return code {out.returncode}")
-    except Exception as error:
-        exception = error
+    exceptions = []
+
+    for attempt in range(max_attempts):
+        try:
+            if not use_cli:
+                s3client.download_file(bucket_name, file.s3file, str(target), Config=config)
+            else:
+                cmd = (
+                    f"aws s3 --endpoint-url {aws_endpoint} --profile {aws_credential} "
+                    f"cp s3://{bucket_name}/{file.s3file}"
+                    f" {target}"
+                ).split()
+                print(cmd)
+                out = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, text=True)
+                if out.returncode != 0:
+                    raise S3DownloadFailed(f"Command {cmd} failed with return code {out.returncode}")
+            # successful download - clear failed errors from previous attempts and break
+            exceptions = []
+            break
+        except Exception as error:
+            exceptions.append(error)
+            if attempt + 1 < max_attempts:
+                print(f"Failed to download file {file}. Retrying - attempt {attempt+2}/{max_attempts}...")
+                time.sleep(1)
+            else:
+                print(f"All {max_attempts} attempts exhausted - failed to download file {file}.")
     elapsed = time.time() - start
     size_gb = file.size / (1024**3)
     print(f"Finished {progress}: {size_gb:.2f}GB in {elapsed:.0f}s ({size_gb/elapsed:.3f} GB/s) for file {target}")
```

Comment:
Might want to modify this log line to better reflect the possible status of the download:
- succeeded after multiple attempts
- Failed

Reply:
I thought it was best not to add more variables and conditions here, so I added logging so that the respective cases (succeed after multiple attempts, and failed) have logging in their respective blocks. And then we only print this final line if it ended in success. What do you think?

Reply:
Yep, that sounds good.